fs-index/src/index.rs

```diff
@@ -1,1120 +1,501 @@
+        path: P,
+    ) -> Option<&IndexedResource<Id>> {
+        self.path_to_resource.get(path.as_ref())
+    }
+
+    /// Build a new index from the given root path
+    pub fn build<P: AsRef<Path>>(root_path: P) -> Result<Self> {
+        log::debug!("Building index at root path: {:?}", root_path.as_ref());
+
+        // Canonicalize the root path
+        let root = fs::canonicalize(&root_path)?;
+        let mut id_to_resources = HashMap::new();
+        let mut path_to_resource = HashMap::new();
+
+        // Loop through the root path and add resources to the index
+        let walker = WalkDir::new(&root)
+            .min_depth(1) // Skip the root directory
+            .into_iter()
+            .filter_entry(should_index); // Skip hidden files
+        for entry in walker {
+            let entry = entry.map_err(|e| {
+                ArklibError::Path(format!("Error walking directory: {}", e))
+            })?;
+            // Ignore directories
+            if !entry.file_type().is_file() {
+                continue;
+            }
+            let path = entry.path();
+            let metadata = fs::metadata(path)?;
+            // Ignore empty files
+            if metadata.len() == 0 {
+                continue;
+            }
+            let last_modified = metadata.modified()?;
+            let id = Id::from_path(path)?;
+            // Path is relative to the root
+            let path = path.strip_prefix(&root).map_err(|_| {
+                ArklibError::Path("Error stripping prefix".to_string())
+            })?;
+
+            // Create the resource and add it to the index
+            let resource = IndexedResource {
+                id: id.clone(),
+                path: path.to_path_buf(),
+                last_modified,
+            };
+            path_to_resource.insert(resource.path.clone(), resource.clone());
+            id_to_resources
+                .entry(id)
+                .or_insert_with(Vec::new)
+                .push(resource);
+        }
 
-            assert_eq!(actual.root, path.clone());
-            assert_eq!(actual.path2id.len(), 0);
-            assert_eq!(actual.id2path.len(), 0);
-            assert_eq!(actual.collisions.len(), 0);
+        Ok(ResourceIndex {
+            root,
+            id_to_resources,
+            path_to_resource,
         })
     }
 
-    #[test]
-    fn should_not_index_1_empty_directory() {
-        run_test_and_clean_up(|path| {
-            create_dir_at(path.clone());
-
-            let actual: ResourceIndex<Crc32> =
-                ResourceIndex::build(path.clone());
+    /// Update the index with the latest information from the file system
+    pub fn update_all(&mut self) -> Result<IndexUpdate<Id>> {
+        log::debug!("Updating index at root path: {:?}", self.root);
+
+        let mut added = Vec::new();
+        let mut modified = Vec::new();
+        let mut removed = Vec::new();
+
+        let new_index = ResourceIndex::build(&self.root)?;
```

Comment:
We should not throw away the previous definition. This implementation is a lot simpler, which is good. But the older implementation was optimized. We should benchmark both functions.

Reply:
I wrote benchmarks to be sure that the proposed logic changes don't impact performance. The `resource_index_benchmark` includes an `index_update_all` function, which:
- Creates 50 files
- Builds the index
- Creates 50 new files, removes 30 files, and modifies 10 files
- Updates the index
I didn't notice any performance drop from the changes proposed in the PR.

This result isn't surprising since the old implementation doesn't seem so different in the core. In the main branch:
- `ResourceIndex::build()` calls `discover_paths()` and `scan_entries()` [here](https://github.com/ARK-Builders/ark-core/blob/main/fs-index/src/index.rs#L53-L54), each once.
- `ResourceIndex::update_all()` calls the same functions plus extra logic:
    - It calls `discover_path()` [here](https://github.com/ARK-Builders/ark-core/blob/main/fs-index/src/index.rs#L207)
    - It calls `scan_entries()` [here](https://github.com/ARK-Builders/ark-core/blob/main/fs-index/src/index.rs#L316) and [here](https://github.com/ARK-Builders/ark-core/blob/main/fs-index/src/index.rs#L320)

Additionally, the code quality improvement is significant imho. The current code for `ResourceIndex::update_all()` is large, error-prone, and difficult to debug. The proposed changes are simpler and easier to debug, with minimal chances of error.

----

To be sure, I pushed similar code to a branch in my fork: [GitHub link](https://github.com/ARK-Builders/ark-core/commit/266347db29e4a899a5cd6a927af9ff1bc0cec527). Benchmarking this against the proposed code gives the following results:
```shell
resource_index/index_update_all
                        time:   [11.218 ms 11.258 ms 11.303 ms]
                        change: [+0.2740% +0.7508% +1.2455%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
```
This shows no noticeable performance hit.


What do you think?

Reply:
I'm not sure that the two definitions are semantically equivalent.

Reply:
Updated the logic for the `update_all()` method. Let me know what you think.

Reply:
Mostly looks good. I suspect there is one bug though (missed negation).

By the way, in benchmarks we should have thousands of files, not dozens.

Reply:
> By the way, in benchmarks we should have thousands of files, not dozens.

For all `ReourceIndex` methods benchmarks or just for `update_all()` ?

Reply:
> By the way, in benchmarks we should have thousands of files, not dozens. 

I agree. just updated the benchmarks temp directory to contain 10,000 files. For `ResourceIndex::update_all()`, we create 5,000 new files, remove 3,000 files, and modify 1,000 files.